.. contents::
   :depth: 3.0
..

References and Resources to Match Poster Sections
=================================================

Background
----------

-  `Nature's special collection: Challenges in Irreproducible
   Research <http://www.nature.com/news/reproducibility-1.17552#/Recent-articles>`__

-  `Full Computational Reproducibility in Biological Science: Methods,
   Software and a Case Study in Protein
   Biology <http://arxiv.org/abs/1608.06897>`__

    "In spite of much recent interest in many scientific areas, emphasis
    remains more on procedures, strictures and discussion, reflecting
    the inexperience of most scientific journals when it comes to
    software..."

-  `Reproducibility, repeatability, and a workshop in
   Norway <http://ivory.idyll.org/blog/2016-norway-repeatability.html>`__

-  `A statistical definition for reproducibility and
   replicability <http://biorxiv.org/content/early/2016/07/29/066803>`__

    "In spite of much recent interest in many scientific areas, emphasis
    remains more on procedures, strictures and discussion, reflecting
    the inexperience of most scientific journals when it comes to
    software..."

-  `Why scientists must share their research code in Nature News
   <http://www.nature.com/news/why-scientists-must-share-their-research-code-1.20504>`__.

-  `Interactive notebooks: Sharing the code by Helen Shen. Nature. 2014
   Nov 6;515(7525):151-2. doi: 10.1038/515151a. PMID:
   25373681 <http://www.nature.com/news/interactive-notebooks-sharing-the-code-1.16261>`__.
   This article is a couple of years old now but gives a good background
   on the issues and impetus for now what are Jupyter Notebooks.

-  `How Jupyter Notebooks Will Improve Your Computational
   Life <http://wp.sanger.ac.uk/barrettgroup/2016/03/10/how-jupyter-notebooks-will-improve-your-computational-life/>`__
   - a nice illustrated introduction to Jupyter Notebooks

-  `Jupyter Notebook Tutorial: The Definitive Guide <https://www.datacamp.com/community/tutorials/tutorial-jupyter-notebook#gs.UeGobjY>`__

-  `An brief introduction to Jupyter Notebook essentials that is meant
   to be hands-on, posted by Titus Brown as part of a recent one day
   seminar <https://2016-oslo-repeatability.readthedocs.io/en/latest/intro-jupyter.html>`__

-  `IPython Notebook/Project Jupyter Half day
   Workshop <http://dib-training.readthedocs.io/en/pub/2016-03-09-jupyter-notebook.html>`__

    "Project Jupyter is a literate data analysis environment (similar to
    knitr and Shiny) that supports over 40 different programming
    languages, including R and Python both. It can be used to build
    reproducible analyses for publication, collaborate over distances on
    data analysis, and build interactive tutorials and homeworks around
    data analysis."

-  `A gallery of interesting IPython
   Notebooks <https://github.com/ipython/ipython/wiki/A-gallery-of-interesting-IPython-Notebooks>`__,
   in particular see `Reproducible academic publications
   section <https://github.com/ipython/ipython/wiki/A-gallery-of-interesting-IPython-Notebooks#reproducible-academic-publications>`__
   . Some other notebooks are highlighted
   `here <http://nbviewer.jupyter.org/>`__.

-  `Advanced Jupyter Notebook Tricks — Part
   I <https://blog.dominodatalab.com/lesser-known-ways-of-using-notebooks/>`__

-  `Pipelining R and Python in
   Notebooks <http://blog.revolutionanalytics.com/2016/01/pipelining-r-python.html>`__

-  `Ten Simple Rules for Effective Statistical
   Practice <http://journals.plos.org/ploscompbiol/article?id=10.1371/journal.pcbi.1004961>`__

    "Modern reproducible research tools like Sweave [18], knitr [19],
    and iPython [20] notebooks take this a step further and combine the
    research report with the code. Reproducible research is itself an
    ongoing area of research and a very important area that we all need
    to pay attention to."

-  Jupyter Notebook will evolve next into a platform that will still run
   in the browser but expose more interface functionality while
   including more, full-featured notebooks, see
   `here <http://blog.jupyter.org/2016/07/14/jupyter-lab-alpha/>`__ to
   see what is planned in the next couple of years as they roll out
   `JupyterLab <http://blog.jupyter.org/2016/07/14/jupyter-lab-alpha/>`__.
   Additionally,
   `JupyterHub <https://jupyterhub.readthedocs.io/en/latest/>`__ is
   being developed to serve multiple persistent, authenticated Jupyter
   Notebooks for teaching and collaborative uses; see slide 21
   `here <http://www.slideshare.net/willingc/jupyterhub-a-thing-explainer-overview>`__
   as a guide to when you may need what implementation.

A sampling of scientific notebooks & extensions
-----------------------------------------------

-  `Developments in next generation
   sequencing <https://flxlexblog.wordpress.com/2016/07/08/developments-in-high-throughput-sequencing-july-2016-edition/>`__
   was plotted using `a
   notebook <https://github.com/lexnederbragt/developments-in-next-generation-sequencing/blob/master/Plotting.ipynb>`__

-  `A collection of Jupyter notebooks authored by the UCSD Center for
   Computational Biology &
   Bioinformatics <https://github.com/ucsd-ccbb/jupyter-genomics>`__

-  `LIGO folks published a Python notebook along with the data to
   explain their analysis and findings (plots, audio
   files). <https://losc.ligo.org/s/events/GW150914/GW150914_tutorial.html>`__
   and a few hours later `you can now use Binder to bring up Jupyter
   with that notebook and all the dependencies preloaded, and step
   through their analysis
   yourself <https://github.com/minrk/ligo-binder>`__

-  `Wang Z and Ma'ayan A. An open RNA-Seq data analysis pipeline
   tutorial with an example of reprocessing data from a recent Zika
   virus study. F1000Research 2016, 5:1574 (doi:
   10.12688/f1000research.9110.1) <https://github.com/MaayanLab/Zika-RNAseq-Pipeline>`__

-  `Bioconductor's RNA-seq Workflow in Jupyter notebook
   format <https://github.com/vladchimescu/rnaseq-workflow/blob/master/rnaseqGene.ipynb>`__
   with related video
   `here <https://www.youtube.com/watch?v=sxn-sixRVtY>`__ with making-of
   described
   `here <https://github.com/vladchimescu/RdocsJupyter/blob/master/vignettes/Intro.Rmd>`__

-  `Multi-tiered genomic analysis of head and neck cancer ties TP53
   mutation to 3p loss, by A. Gross et al. (Nature Genetics 2014) with
   he full collection of notebooks to replicate the
   results <https://github.com/theandygross/TCGA/tree/master/Analysis_Notebooks#guide-to-running>`__

-  `Exploratory bioinformatics with plot.ly and IPython notebook:
   Visualizing gene expression
   data <https://plot.ly/ipython-notebooks/bioinformatics/>`__ features
   using a high-end plotting interface with bioinformatics data

-  `Introduction to Applied Bioinformatics (or IAB) is a free, open
   source interactive text that introduces readers to core concepts of
   bioinformatics in the context of their implementation and
   application. <http://readiab.org/>`__ - uses Jupyter Notebooks and
   mybinder.org

-  `Exploring proteomics data from TCGA/CPTAC breast cancer
   samples <https://github.com/hussius/TCGA_proteomics_tutorial/blob/master/TCGA_protein_tutorial.ipynb>`__
   as described
   `here <https://followthedata.wordpress.com/2016/02/09/tutorial-exploring-tcga-breast-cancer-proteomics-data/>`__

-  `Dynamics and associations of microbial community types across the
   human body <http://dx.doi.org/10.1038/nature13178>`__, by Tao Ding &
   Patrick D. Schloss. `Notebook replicating
   results <http://nbviewer.ipython.org/gist/pschloss/9815766/notebook.ipynb>`__

-  `Indication of family-specific DNA methylation patterns in developing
   oysters <http://biorxiv.org/content/early/2014/12/16/012831>`__,
   Claire E. Olson, Steven B. Roberts doi:
   http://dx.doi.org/10.1101/012831. `Notebook to generate results in
   the
   paper <http://nbviewer.ipython.org/github/che625/olson-ms-nb/blob/master/BiGo_dev.ipynb>`__.

-  `Transcriptome Sequencing Reveals Potential Mechanism of Cryptic 3’
   Splice Site Selection in *SF3B1*-mutated
   Cancers <http://journals.plos.org/ploscompbiol/article?id=10.1371/journal.pcbi.1004105>`__
   by `Christopher DeBoever <http://cdeboever3.github.io/>`__ et al.
   There are several
   `notebooks <https://github.com/cdeboever3/deboever-sf3b1-2015/tree/master/notebooks>`__
   to replicate results and make figures.

-  `A Reference-Free Algorithm for Computational Normalization of
   Shotgun Sequencing Data, by C.T. Brown et
   al. <http://ged.msu.edu/papers/2012-diginorm>`__. `Full
   notebook <http://nbviewer.ipython.org/urls/github.com/ged-lab/2012-paper-diginorm/raw/master/notebook/diginorm.ipynb>`__

-  The Broad Institute built `an extension for working their GenePattern
   platform from within a Jupyter Notebook
   environment <http://software.broadinstitute.org/cancer/software/genepattern/genepattern-notebook-for-jupyter-users>`__

    Be sure to look at `the list of example notebooks using the
    Github/Binder
    approach <http://retreat16.readthedocs.io/en/latest/references%20by%20section/#other-noteable-notebooks-using-github-binder-approach>`__
    below as well.

Resources for Running Active Notebooks in the Cloud
---------------------------------------------------

-  `Freeman Lab's MyBinder.org site <http://mybinder.org>`__ - where
   you'll go to point their system at your Github repository with a
   Jupyer Notebook to make active notebooks available online

-  `Is mybinder 95% of the way to next-gen computational science
   publishing, or only
   90%? <http://ivory.idyll.org/blog/2016-mybinder.html>`__

    "The split that my lab has made here is to use a workflow engine
    (e.g. make, pydoit, or snakemake) for the compute & data intensive
    stuff, and then feed those intermediate results (assembly and
    mapping stats, quantification, etc.) into analysis notebooks. For
    mybinder purposes, there should be no problem saving those
    intermediate results into a github repo for us and everyone else to
    analyze and reanalyze."

-  `tmpnb.org <http://tmpnb.org>`__ or
   `try.jupyter.org <https://try.jupyter.org/>`__ - launch active,
   transient Jupyter Notebooks in the cloud for basic development, see
   `Instant Temporary IPython
   Notebooks <https://lambdaops.com/ipythonjupyter-tmpnb-debuts/>`__

I have made `a
page <http://retreat16.readthedocs.io/en/latest/using%20tmpnb/>`__ to
walk you through trying tmpnb.org or try.jupyter.org, find it
`here <http://retreat16.readthedocs.io/en/latest/using%20tmpnb/>`__.

-  Another alternative is to click
   `here <https://app.dominodatalab.com/nick/py3-demo>`__. I have not
   extensively run this site through it's paces, and so I can offer a
   couple points about it and beyond that your mileage may vary. It was
   set up by the excellent Domino Data Lab to serve as a place to run
   `an active notebook about differences between Python 2.x and
   3.x <https://blog.dominodatalab.com/adding-python-3-to-jupyter/>`__
   without need for signing into the Domino Data Lab service. I know the
   free tier for signed in users is limited to 15 minutes, and so maybe
   the anonymous one also has this limitation as well? Also similar to
   tmpnb.org unless it is already installed, you won't have access to
   other modules or be able to scrape data from other sites.

The Binder/Github set-up allows you to designate other modules you need
loaded when the instance is spun-up, but that does mean some set-up
steps as discussed in `the
appendix <http://retreat16.readthedocs.io/en/latest/making%20binder/>`__.
Contact me if you need help understanding how to set this up.

Launching Active Versions of My Notebooks
-----------------------------------------

Click the |Binder| button at any of the following repositories for an
active notebook:

`Ammonium Sulfate Precipitation Screen
Calculator <https://github.com/fomightez/small_scale_ammonium_sulfate_precipitation_calculator>`__

`Cell Density
Estimator <https://github.com/fomightez/methods_in_yeast_genetics/blob/master/cell_density_estimator/>`__

The Cell Density Estimator where `only a single sample is
analyzed <http://nbviewer.jupyter.org/github/fomightez/methods_in_yeast_genetics/blob/master/cell_density_estimator/cell_density_estimator_for_single_sample.ipynb>`__
may be easier for novices to follow, and an active notebook can be
launched
`here <http://mybinder.org/repo/fomightez/methods_in_yeast_genetics/notebooks/cell_density_estimator/cell_density_estimator_for_single_sample.ipynb>`__.

`Yeast Growth
Planner <https://github.com/fomightez/methods_in_yeast_genetics/blob/master/yeast_growth_planner/>`__

`Notebook designed as an active computing exercise for young students
visiting the lab <https://github.com/fomightez/uscad16>`__

The example notebook used for the introduction section of the poster can
be found `here <https://github.com/fomightez/jupyter_nb_basics>`__

Contrast the transparency of the
``Ammonium Sulfate Precipitation Screen Calculator`` with a form-based
Django site that performs same calculation
`here <http://fomightez.pythonanywhere.com/ammonium_screen/>`__.

Other Noteable Notebooks Using Github/Binder approach
-----------------------------------------------------

-  `Molecular Design Toolkit
   Demo <https://github.com/avirshup/mdt-gallery-test>`__ - after
   hitting the |Binder| button there I suggest the early parts of the
   ``Example 1. Build and simulate DNA.ipynb`` and
   ``Example 3. Simulating a crystal structure.ipynb`` notebooks.

-  `nglview <https://github.com/arose/nglview>`__ is a Python package
   that makes it easy to visualize molecular systems, including
   trajectories, directly in the Jupyter Notebook. (Launch a Binder by
   clicking the ``Binder`` logo there.) See more about nglview
   `here <http://www.mdanalysis.org/2016/03/14/nglview/>`__.

-  `VPython - Visual Python
   demos <https://github.com/BruceSherwood/vpython-jupyter>`__ has a
   |Binder| button at the bottom. Try ``Atomic solid`` for a simulation
   of interatomic interactions.

-  `Introduction to Applied Bioinformatics (or IAB) is a free, open
   source interactive text that introduces readers to core concepts of
   bioinformatics in the context of their implementation and
   application. <http://readiab.org/>`__

-  \*\* the LIGO notebook is most famous Jupyter Notebook presently, and
   it is available in active form\*\* - LIGO folks published a Python
   notebook along with the data to explain their analysis and findings
   of gravitational waves and `you can now use Binder to bring up
   Jupyter with that notebook and all the dependencies preloaded, and
   step through their analysis
   yourself <https://github.com/minrk/ligo-binder>`__

.. |Binder| image:: http://mybinder.org/badge.svg
